### DISCUSSION TITLE
Attempting to install in docker container but keep running into "bash: /usr/local/bin/docker-compose: cannot execute binary file: Exec format error"

### DISCUSSION THREAD

**Discussion by Inkertus** [alexjustesen/speedtest-tracker#593] [Q&A]
I appear to be doing everything according to the videos I have seen thus far, but keep running into that pesky exec format error on my Pi 4B. There does appear to be a version coded for ARM but I'm lost at this point. 
Any help?

**Comment by alexjustesen** [comment]
Are you using 32bit or 64bit version of PI OS (or something else) on your 4B?

**Reply by Inkertus** [reply]
pi-serv-user@raspberrypi:~ $ uname -m
aarch64

(Raspberry Pi OS most recent version and updates)

**Reply by alexjustesen** [reply]
I'll take a look but rpi issues have seemingly been pretty random when it comes to Docker but check these items...

1. Don't specify a version so the latest is pulled
2. Make sure `docker-compose` is installed
3. Different platform versions are published so you can always try a specific build

I have a 4B (8GB) at home that this runs on fine so I don't think it's the images being built.

**Reply by Inkertus** [reply]
I must not know enough what I'm doing because I'm trying to follow the Docker Compose bit and it keeps throwing errors. Times like this make me wish for basic, easy-to-use "double-click to run" stuff.

**Reply by Inkertus** [reply]
and yes I do have docker compose installed

**Reply by alexjustesen** [reply]
Can you share the composer script you're using or are you copying it from the docs as is?

**Reply by Inkertus** [reply]
![image](https://github.com/alexjustesen/speedtest-tracker/assets/39816686/66e9535a-6f22-42c5-ab55-200cd3f89087)

This, or a script 99% like it, is the one I've been trying to use. I have been known to be blind occasionally but I don't believe this to be the case here.

**Reply by alexdelprete** [reply]
the `exec format error` you were having is due to the fact you are running `docker-compose` instead of `docker compose`.

you have probably an old installation of docker/docker-compose on the system.

run `docker version` and `docker compose version` to check what versions you have installed.

**Comment by alexdelprete** [comment]
are you really using `/path/to/directory` in the compose file?

That has to be a real directory...it was just an example in the docs to tell the user to configure the appropriate path.

**Reply by Inkertus** [reply]
I barely know what I am doing here, I fully admit that. I know enough to follow a guide and do basic troubleshooting for command line and scripting projects and usually wander my way thru. I have edited the file to 
![Screenshot 2023-05-13 071753](https://github.com/alexjustesen/speedtest-tracker/assets/39816686/4a198fb1-4a19-40a4-8606-e972de2e85b2)
and also tried 
 volumes:
            - '/home/pi-serv-user/docker/speedtest-tracker/docker-compose.yml:/config'
       
with the same error.

**Reply by Inkertus** [reply]
...in accordance with the tutorial I tried following.

**Comment by alexdelprete** [comment]
@Inkertus if you use docker, you have to know the MINIMUM stuff, the basics, otherwise you will always have issues.

`/home/pi-serv-user/docker/speedtest-tracker/docker-compose.yml:/config`

you are mapping a file to a folder, I don't know what guide you are following, but you are not reading the docs.

I'll help you, but please read something about docker and docker-compose in the future.

I assume the folder `~/docker/speedtest-tracker` in which you have the docker-compose file is a specific folder for speedtest-tracker you created. If so, create two subfolders in there `config` and `certs`:

This is the folder that contains `docker-compose.yml` file:
- `~/docker/speedtest-tracker/docker-compose.yml` 

These are the subfolders that need to be present:
- `~/docker/speedtest-tracker/config`
- `~/docker/speedtest-tracker/certs`

Then modify `~/docker/speedtest-tracker/docker-compose.yml` like this:
```
version: '3.3'
services:
  speedtest-tracker:
    image: ghcr.io/alexjustesen/speedtest-tracker:latest
    container_name: speedtest-tracker
    restart: unless-stopped
    ports:
      - 8080:80
      - 8443:443
    environment:
      - PUID=1000
      - PGID=1000
      - SSL_MODE=mixed
    volumes:
      - $PWD/config:/config
      - $PWD/certs:/etc/ssl/web
```

`$PWD` will be expanded to the folder where the docker-compose is (~/docker/speedtest-tracker)

After you updated the docker-compose file, run this to recreate the container:

`docker compose up -d --force-recreate --remove-orphans`

**Reply by Inkertus** [reply]
Thank you for your help so far. I have modified as instructed, with and without the last line in the document. The IDs are 1000 for both.
![Screenshot 2023-05-13 071753](https://github.com/alexjustesen/speedtest-tracker/assets/39816686/62d96761-4ca6-40f5-bc99-78b6830c52f2)

**Reply by Inkertus** [reply]
I got rid of the SSL_MODE=mixed and ran again. Now it seems to be running

**Comment by alexdelprete** [comment]
One more thing:

PUID and PGID have to be mapped to the user owning the config and certs subfolders. You are using 1000, but you need to verify what is the ID of the user you are running (pi-serv-user).

So run this command: `id pi-serv-user`

You will have something like this (my user is `alexdelprete`):

```
❯ id alexdelprete
uid=1000(alexdelprete) gid=1000(alexdelprete)
```

The numbers you will see (in my case it's 1000 and 1000) are the ones you need to use in the docker-compose file as PUID and PGID, ok?

**Comment by alexdelprete** [comment]
Please don't post screenshots, copy&paste the output of the shell in the future, using <> symbol in the editing bar to format the text as code.

There was an error, my bad, I updated the environment section in the docker-compose file, like this:

```
    environment:
      - PUID=1000
      - PGID=1000
      - SSL_MODE=mixed
```

Update docker-compose.yml and run: `docker compose up -d --force-recreate --remove-orphans`

**Comment by Inkertus** [comment]
Again, thank you very much. I'm headed out for now but I did get it running and a non-default user configured. I do need to get the documentation up and read so I can at least get an idea of what I'm doing and be less of (I figure) a bother. 
Thanks again for both the help and the project.

**Reply by alexdelprete** [reply]
No problem @Inkertus, no bother. Just my recommendation to all users to have a minimum understanding of what they're using so they can troubleshoot and maintain the services they're using. It's not setup & install like on windows. :)

Enjoy ST.
